### PR TITLE
feat: Generate DynamoDB resource policies

### PR DESCRIPTION
**Pull request by skuenzli**
Add support for generating k9-style least privilege resource policies for DynamoDB resources.

```
const ddbResourcePolicyProps: k9.dynamodb.K9DynamoDBResourcePolicyProps = {
    k9DesiredAccess: new Array<k9.k9policy.IAccessSpec>(
        {
        ... desired access capabilities ...
        note: you should explicitly-allow your AWS Access Analyzer service-linked role to read-config
              so that you don't get an Error finding in AWS AA because it can't read the table metadata & policy
        });
const ddbResourcePolicy = k9.dynamodb.grantAccessViaResourcePolicy(ddbResourcePolicyProps);

const table = new dynamodb.TableV2(stack, 'k9-cdk-v2-int-test', {
  partitionKey: { name: 'pk', type: dynamodb.AttributeType.STRING },
  removalPolicy: cdk.RemovalPolicy.DESTROY,
  resourcePolicy: ddbResourcePolicy
});
```

### CODE REVIEW

File: resources/capability_summary.json [hunk 1]
```diff
@@ -70,70 +70,49 @@
   "DynamoDB": {
     "administer-resource": [
       "dynamodb:CreateBackup",
-      "dynamodb:CreateGlobalTable",
```
Comment (joshlartz) [inline resources/capability_summary.json:73]: These permissions for global tables very likely need to stay. The admin user should be allowed to make replicas.

File: src/dynamodb.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+export const SID_DENY_EVERYONE_ELSE = 'DenyEveryoneElse';
+
+/**
+ * Generate a DynamoDB resource policy from the provided props that can be attached to DynamoDB
+ * resources, particularly tables & indices.
+ *
+ * @param props specifying desired access
+ * @return a PolicyDocument that can be attached to DynamoDB resources
+ */
+export function makeResourcePolicy(props: K9DynamoDBResourcePolicyProps): PolicyDocument {
```
Comment (joshlartz) [inline src/dynamodb.ts:33]: This works for our needs, but I was wondering if you still wanted to wrap this in a `grantAccessViaResourcePolicy` method to be consistent with s3.

File: src/dynamodb.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  let accessSpecsByCapabilityRecs = policyFactory.mergeDesiredAccessSpecsByCapability(SUPPORTED_CAPABILITIES, props.k9DesiredAccess);
+  let accessSpecsByCapability: Map<AccessCapability, IAccessSpec> = new Map();
+
+  for (let [capabilityStr, accessSpec] of Object.entries(accessSpecsByCapabilityRecs)) {
+    accessSpecsByCapability.set(getAccessCapabilityFromValue(capabilityStr), accessSpec);
+  }
+
+  if (!canPrincipalsManageResources(accessSpecsByCapability)) {
+    throw Error('At least one principal must be able to administer and read-config for DynamoDB resources' +
```
Comment (joshlartz) [inline src/dynamodb.ts:47]: Good check to have.

File: src/k9policy.ts [hunk 4]
```diff
@@ -61,6 +61,44 @@ export interface IAWSServiceAccessGenerator {
...
+    const readConfigPrincipals = new Set<string>(readConfigSpec.allowPrincipalArns);
+    const intersection = new Set(
+      [...adminPrincipals].filter(x => readConfigPrincipals.has(x)));
+    return intersection.size > 0;
+  }
+  return false;
+}
+
+
+export function toPascalCase(input: string): string {
```
Comment (joshlartz) [inline src/k9policy.ts:92]: Did you hit a different naming restriction here that forced the casing?

File: test/k9.test.ts [hunk 5]
```diff
@@ -534,6 +536,84 @@ describe('K9KeyPolicy', () => {
 
 });
 
+describe('DynamoDBResourcePolicy', () => {
```
Comment (joshlartz) [inline test/k9.test.ts:539]: nit: break the tests into files to match src.

File: src/k9policy.ts [hunk 6]
```diff
@@ -61,6 +61,44 @@ export interface IAWSServiceAccessGenerator {
...
 }
 
+/**
+ * Check whether the provided access specs ensure that at least one principal can both read and administer configuration.
+ * @param accessSpecsByCapability is a map of access specs keyed by access capability
+ *
+ * @return true when at least one principal that can administer and read configuration exists
+ */
+export function canPrincipalsManageResources(accessSpecsByCapability: Map<AccessCapability, IAccessSpec>) {
+  console.log(`canPrincipalsManageResources eval'ing ${accessSpecsByCapability}`);
```
Comment (joshlartz) [inline src/k9policy.ts]: Console logging in a CDK library is a bit unusual. I'd advise debug here. 

I forgot to mention that a while back. I swallow your console.log's from the KMS key policy generation to hide them. :)

File: src/k9policy.ts [hunk 7]
```diff
@@ -61,6 +61,44 @@ export interface IAWSServiceAccessGenerator {
...
+  let adminSpec = accessSpecsByCapability.get(AccessCapability.ADMINISTER_RESOURCE);
+  let readConfigSpec = accessSpecsByCapability.get(AccessCapability.READ_CONFIG);
+
+  if ((adminSpec?.allowPrincipalArns && adminSpec.allowPrincipalArns.length > 0)
+        && (readConfigSpec?.allowPrincipalArns && readConfigSpec.allowPrincipalArns.length > 0)) {
+    const adminPrincipals = new Set<string>(adminSpec.allowPrincipalArns);
+    const readConfigPrincipals = new Set<string>(readConfigSpec.allowPrincipalArns);
+    const intersection = new Set(
```
Comment (joshlartz) [inline src/k9policy.ts:78]: `Set` has a native intersection method.

File: test/__snapshots__/k9.test.ts.snap [hunk 8]
```diff
@@ -1,5 +1,213 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`DynamoDBResourcePolicy Typical usage 1`] = `
+Object {
```
Comment (joshlartz) [inline test/__snapshots__/k9.test.ts.snap:4]: Jest 28 or 29 finally improved snapshot syntax and removed the type literals everywhere. Much easier on the eyes.

**Review by joshlartz — APPROVED** [review]

File: resources/capability_summary.json (hunk 1 above)
Comment (skuenzli) [reply resources/capability_summary.json:73]: These permissions changes were the result of:
1. updating to the full current list of DDB permissions
2. removing permissions not [supported by DDB resource policy](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/rbac-iam-actions.html)

Here are the `administer-resource` permissions from the full set of DDB perms that the DDB resource policy API rejected:
```
(administer-resource) One or more parameter values were invalid: Invalid policy document: The following action names are invalid: "dynamodb:UpdateGlobalTable", "dynamodb:UpdateGlobalTableSettings", "dynamodb:CreateTable", "dynamodb:UpdateGlobalTableVersion", "dynamodb:CreateGlobalTable", "dynamodb:RestoreTableFromBackup", "dynamodb:RestoreTableFromAwsBackup", "dynamodb:CreateTableReplica", "dynamodb:PurchaseReservedCapacityOfferings", "dynamodb:ImportTable"
```

(I have similar output for `read-config`, `read-data`, and `write-data` if you are interested)

I can see why you logically you want an admin of a table to create a replica. However, I think that probably has to be granted from the Identity side. That aligns more with how `Create[Resource]` is generally not supported in resource policy. And in this specific case, I would think the principal CFn uses to create the table resources would be the one that needs permissions.

wdyt?

File: src/k9policy.ts (hunk 4 above)
Comment (skuenzli) [reply src/k9policy.ts:92]: Correct. DDB doesn't allow spaces nor hyphens.

File: test/k9.test.ts (hunk 5 above)
Comment (skuenzli) [reply test/k9.test.ts:539]: yes, might do that.

File: src/k9policy.ts (hunk 6 above)
Comment (skuenzli) [reply src/k9policy.ts]: sorry, that's not supposed to be there. will remove.

**Comment by skuenzli** [comment]
For future reference...

The DynamoDB permissions changes were the result of:

1. updating to the full current list of DDB permissions c.f. [the k9 access capabilities API](https://docs.k9security.io/API/get-access-capabilities/)
2. removing permissions not [supported by DDB resource policy](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/rbac-iam-actions.html)

Here are from full set of DDB perms that the DDB resource policy API rejected:

```
One or more parameter values were invalid: Invalid policy document: The following action names are invalid:

(administer-resource): "dynamodb:UpdateGlobalTable", "dynamodb:UpdateGlobalTableSettings", "dynamodb:CreateTable", "dynamodb:UpdateGlobalTableVersion", "dynamodb:CreateGlobalTable", "dynamodb:RestoreTableFromBackup", "dynamodb:RestoreTableFromAwsBackup", "dynamodb:CreateTableReplica", "dynamodb:PurchaseReservedCapacityOfferings", "dynamodb:ImportTable"

(read-config): "dynamodb:ListContributorInsights", "dynamodb:ListGlobalTables", "dynamodb:DescribeGlobalTableSettings", "dynamodb:DescribeReservedCapacityOfferings", "dynamodb:ListBackups", "dynamodb:DescribeLimits", "dynamodb:ListExports", "dynamodb:DescribeEndpoints", "dynamodb:DescribeReservedCapacity", "dynamodb:ListImports", "dynamodb:ListTables", "dynamodb:DescribeStream", "dynamodb:DescribeImport", "dynamodb:DescribeGlobalTable", "dynamodb:DescribeBackup", "dynamodb:ListStreams"

(read-data): "dynamodb:GetRecords", "dynamodb:GetShardIterator"

(write-data): "dynamodb:RestoreTableFromAwsBackup", "dynamodb:StartAwsBackupJob", "dynamodb:CreateTableReplica", "dynamodb:ImportTable"

(delete-data): "dynamodb:DeleteBackup"
```

File: src/k9policy.ts (hunk 7 above)
Comment (skuenzli) [reply src/k9policy.ts:78]: Unfortunately, I don't see an `intersection` method available on TypeScript `5.4.5` & Node `20.14.0` (what I dev on).

Looks like an upgrade might help here.  However, it's tough when CDK supports all the way back to Node `14.15.0` and I know we have users on `16.x`

**Pull request merged**